### PR TITLE
Fix broken diff search if sub-repo perms are enabled

### PR DESCRIPTION
**Pull request by pjlast**
When sub-repo permissions are enabled, the sub-repo permissions filter would be applied to repositories where sub-repo permissions are relevant. This lead to `type:diff` search being broken when sub-repo permissions are enabled, since it would try to filter repositories for which sub-repo permissions do not apply.

This PR changes the check to confirm that sub-repo permissions are enabled for the repository being filtered before doing the sub-repo filter, and if the repository does not have sub-repo permissions enabled, the repository is skipped.

## Test plan

Verified with manual checks. Added unit test.

### CODE REVIEW

**Review by sashaostrikov — APPROVED** [review]
Nice fix!

**Review by erzhtor — APPROVED** [review]

**Review by kopancek — APPROVED** [review]
Do we want to backport this @pjlast ?

File: CHANGELOG.md [hunk 1]
```diff
@@ -64,6 +64,7 @@ All notable changes to Sourcegraph are documented in this file.
 - SAML assertions to get user display name are now compared case insensitively and we do not always return an error. [#52992](https://github.com/sourcegraph/sourcegraph/pull/52992)
 - The braindot menu on the blob view no longer fetches data eagerly to prevent performance issues for larger monorepo users. [#53039](https://github.com/sourcegraph/sourcegraph/pull/53039)
 - Fixed an issue where commenting out redacted site-config secrets would re-add the secrets. [#53152](https://github.com/sourcegraph/sourcegraph/pull/53152)
+- Fixed an issue where `type:diff` search would not work when sub-repo permissions are enabeld. [#53210](https://github.com/sourcegraph/sourcegraph/pull/53210)
```
Comment (keegancsmith) [inline CHANGELOG.md:67]: typo: enabled

File: internal/search/job/jobutil/sub_repo_perms_job.go [hunk 2]
```diff
@@ -85,6 +86,15 @@ func applySubRepoFiltering(ctx context.Context, checker authz.SubRepoPermissionC
 	filtered := matches[:0]
 
 	for _, m := range matches {
+		enabled, err := authz.SubRepoEnabledForRepo(ctx, checker, m.Key().Repo)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Could not determine if sub-repo permissions are enabled for repo %s. Skipping.", m.Key().Repo))
```
Comment (keegancsmith) [inline internal/search/job/jobutil/sub_repo_perms_job.go:91]: this is not how you should use logger. Pass a message then use `log.String("repo", m.Key().Repo)` as a parameter

**Review by keegancsmith — COMMENTED** [review]
It is surprising to me that authz.CanReadAnyPath doesn't return true if subrepoperms are disabled. That feels like the proper fix?

**Comment by keegancsmith** [comment]
Fixes https://github.com/sourcegraph/sourcegraph/issues/53190

File: internal/search/job/jobutil/sub_repo_perms_job.go (hunk 2 above)
Comment (pjlast) [reply internal/search/job/jobutil/sub_repo_perms_job.go:91]: Woops, let me make a follow-up PR. Thanks!

**Pull request merged**